### ISSUE TITLE
Rethink - and update database  -  with new name for review_assignment

### ISSUE THREAD

**Issue by nmadruga** [openmsupply/application-manager-server#189]
Issue created based on discussion on the planning meeting of Sprint 18 (22/01/2021):
* Each Consolidator (lvl > 1) with permission to consolidate an application will get one review_assignment created once Application is Submitted.
* When the first Review (lvl 1) is Submitted all review_assignment (of lvl > 1) will get a review created, the status should be `pending` until one starts the consolidation. Then the current consolidation review status changes to `draft` and the others should get status of `locked` forever.
* Similar thing to happen when consolidation (lvl 2) finishes and other review_assignment (lvl > 2) get reviews created for them.

All this needs to be:
1. Decided - there is still debate weather this should be the proposed workflow or if we shouldn't create reviews until one consolidation has started (as we do with review lvl 1).
2. Documented
3. Be cleared out in the schema by renaming the `review_assignment` and all calls from the front-end to this field.

**Comment by CarlosNZ** [comment]
Okay, so I think the main issue I had with the proposed assignment/new review design comes down to the fact that the `review_assignment` table is not actually being used for what it says it is, and in fact is serving two separate purposes in two different situations (initial review / consolidation).

In initial review, it’s an “assignment” record — which reviewer is doing what. However, for consolidation, it’s no longer being used to show assignment as such, but as a pre-computed table of which reviewers are _available_ for consolidating/reviewing tasks on different applications. Now that we’re combining reviews and consolidations into one general “review” structure, I think it’s very confusing to have the same table serve different purposes in different situations and will lead to unnecessary complexity and mistakes down the line.

Just to be clear, I’m completely on board with the idea of “pre-computing” available reviewers for each application, for the purposes of simplifying the filtering/query logic in the front-end. That makes sense. But I think we should be explicit about that, and do it consistently, so it’s clear from the schema what is going on and doesn’t mean two different things at different times.

I have an idea for how this might work, would be good to get your feedback about potential oversights, problems, etc. and improvements.

I’ve attached a diagram, but here are the main points:

- rename `review_assignment` table to `review_availability` to better reflect what it’s actually doing
- an additional field in this table called `is_assigned`, which is `NULL` by default, marked `true` when a review is _actually_ assigned, and `false` when no longer available (i.e when another consolidator starts/self-assigns the application/stage -- equivalent to "locked")
- do the same thing for initial review and consolidation, the only difference is that initial review is assigned by a supervisor whereas consolidations are (usually) self-assigned.
- the task of updating the `review_availability` is done by a back-end Action. (probably runs `onReviewSubmit`, `onApplicationSubmit`, and `onAssignment` -- TBD)
- new reviews/consolidations are not created until the `is_assigned` field is marked `true`.  (To avoid all the unnecessary “dummy” records, which just add confusion IMO). Not sure if we create the record literally _on_ assignment, or when the review chooses to _start_, but I don’t think this really affects anything.
- the main point of difficulty as far as I can tell is that first level review can have more than one assignment (as in a different reviewer per section), whereas consolidations don’t. One way to handle this would be: the Action that runs when someone is assigned (i.e. `is_assigned` marked true) checks if all _responses_ have been assigned and, if so, marked all other availability entries `false`, else leaves them `null`.  Could probably do with a bit more clarity/discussion on this point, but I think it should be able to work okay.

Note that this suggestion only involves the whole review/assignment level. All the existing ideas for handling responses (duplication, draft, etc.) seem great as is.

Anyway, the diagram: (original [here](https://drive.google.com/open?id=11811CN1_GginqJT_QlFDPVWvdP0Si8vw&authuser=carl%40msupply.foundation&usp=drive_fs))

![Review Availability 2021-01-26](https://user-images.githubusercontent.com/5456533/105903847-446f6500-6085-11eb-8a2e-5670f4b8cda1.png)

**Comment by nmadruga** [comment]
Thanks for having a think about this one. There are many of these in-conclusions that we need to make sure we all agree before we start implementing this. Otherwise it would be very hard to understand and keep all very well connected.

From my start reading this I have one question about this:
> do the same thing for initial review and consolidation, the only difference is that initial review is assigned by a supervisor whereas consolidations are (usually) self-assigned.

Do you meant to create one `review_availability` for every reviewer (lvl 1) even before they get assigned to a section?

**Comment by CarlosNZ** [comment]
> Do you meant to create one `review_availability` for every reviewer (lvl 1) even before they get assigned to a section?

Yes, I think so. Both for consistency (the table does one thing), and it would provide the same benefit as it's doing for consolidation -- a pre-computed list of who's available (for assigners) to simplify front-end querying.

**Comment by nmadruga** [comment]
Cool. Just wanted to confirm that.

Also another thing about this one:

> new reviews/consolidations are not created until the is_assigned field is marked true. (To avoid all the unnecessary “dummy” records, which just add confusion IMO). Not sure if we create the record literally on assignment, or when the review chooses to start, but I don’t think this really affects anything.

One problem that we were trying to avoid (by creating the consolidation review on the go) was not to let this be created by front-end mutation. So we wouldn't need to add the permission to `create` on the user with consolidation permission. Although now that we have combined with the **Review** table, it seems to be ok to create this with a mutation (since this is already done for review lvl 1).

**Comment by andreievg** [comment]
Hi, I am buying it so far. A few questions and suggestions though:

> an additional field in this table called is_assigned, which is NULL by default, marked true when a review is actually assigned, and false when no longer available (i.e when another consolidator starts/self-assigns the

Maybe at this stage `is_assigned` should be turned ENUM ? I've realised we've used already used nulls in a few places so OK to keep, but would make schema more readable/understandable if we avoid using NULL to represent things. (what would be a good ENUM values and field name to represent, "not assigned but triaged", "assigned", "self-assigned", "someone else self-assigned or application fully assigned". 

Say this field was status (and schema would clearly show that review_assignment has different statuses), could we keep `review_assignment` naming ?

![Screen Shot 2021-01-27 at 11 24 27 AM](https://user-images.githubusercontent.com/26194949/105913805-44c22d00-6092-11eb-809b-1bd5da52ddc7.png)

#### On Application (FIRST) Submitted.

Do we create review_availability records for all possible reviewers, and the assigner then able to assign sections/questions to review (by linking review_question_assignment) ?

I wonder why we would need this though ? I mean what problem does it solve ? VS for lvl 1 just letting assigner create review_availability records (since reviewers will have restrictions 'by section' etc.. so assigner would need to probably query an end point to get a list of reviewers and sections they can review, thus don't see a need of pre made review_assignments).

#### On Review Submitted

So this logic would be, if `consolidation does not exist` and if review_availability records don't exist we create review_availability records ? If `consolidation exists` we change status to pending ?

I think this would work nicely (would set assigner_id to reviewer_id at this stage, to make it easer for permissions to restrict editing of assignments)

![Screen Shot 2021-01-27 at 11 46 12 AM](https://user-images.githubusercontent.com/26194949/105916039-46412480-6095-11eb-935d-04da5dc9985a.png)

This looks good, although you mention 'actions', I would assume that you mean front end:
* mutation to self assign (triggers an action to set other review_availabity to "someone else self-assigned" )
then
* mutation to create review (by default would have 'draft')
* mutation to create responses.

![Screen Shot 2021-01-27 at 11 55 23 AM](https://user-images.githubusercontent.com/26194949/105916910-8f45a880-6096-11eb-897f-12b75c7643f6.png)

##### For Assigners

This links a little bit to the comments above. How does the assigner know that all of the application has been assigned (just by looking at review_availability ?).

We haven't quite narrowed in on this (in previous design as-well). But the idea is to have some view or computed column for this.

There is a mention of `level`, what do you mean by that ?

There is a bit of extra logic that needs to go to  /review and /consolidation `url` where previously we just find the latest review, now we would need to find the latest review or review_availabity for the current stage. So we would need to amend that diagram.

**Comment by CarlosNZ** [comment]
> Maybe at this stage `is_assigned` should be turned ENUM ? I've realised we've used already used nulls in a few places so OK to keep, but would make schema more readable/understandable if we avoid using NULL to represent things. (what would be a good ENUM values and field name to represent, "not assigned but triaged", "assigned", "self-assigned", "someone else self-assigned or application fully assigned".

Yeah, fair enough, we're starting to slip into using NULL as a third option in a boolean a bit too often ;) So yeah, an enum with values: "available", "assigned", "locked" perhaps? (We don't need a separate "self-assigned" value I think, cause in that case `assigner_id` would be the same as `reviewer_id` -- that's how you'd tell it was self-assigned.)

> Say this field was status (and schema would clearly show that review_assignment has different statuses), could we keep review_assignment naming ?

So keep the table named `review_assignment`, and this column called `status` with the above values? Yeah, that works, it clearly shows the difference between "available" and "assigned", which is the key thing I think.

> Do we create review_availability records for all possible reviewers, and the assigner then able to assign sections/questions to review (by linking review_question_assignment) ?

Yes, that's what I was imagining.

> I wonder why we would need this though ? I mean what problem does it solve ? VS for lvl 1 just letting assigner create review_availability records (since reviewers will have restrictions 'by section' etc.. so assigner would need to probably query an end point to get a list of reviewers and sections they can review, thus don't see a need of pre made review_assignments).

Yeah, I see your point about needing to look up Section permissions anyway. It would still probably help with front-end filtering/querying though -- if the assigner wants their list to show "unassigned applications", or "applications needing review assignments", this table would probably simplify that (not 100% sure, need to talk that bit through some more). Actually the `review_assignment/availability` table can probably account for Section restrictions -- when it re-computes values (on Assignment), a reviewer who no longer has sections that they are allowed to review, would have their review_assignment record set to "locked" (or whatever the inactive label is).

> So this logic would be, if consolidation does not exist and if review_availability records don't exist we create review_availability records ? If consolidation exists we change status to pending ?

Yes, I think that sounds right. Whatever it would need to be consistent with your existing process in that regard.

> This looks good, although you mention 'actions', I would assume that you mean front end:
> -  mutation to self assign (triggers an action to set other review_availabity to "someone else self-assigned" )
>    then
> -  mutation to create review (by default would have 'draft')
> -  mutation to create responses.

I've tried to say "Action" when i mean back-end Action, and "actionable" when I mean front-end "thing that user can do". Hopefully I've been consistent with that, let me know if not.

Thinking about it more, I think the simplest option is that, for both level 1 reviewers and consolidators, the review record is not created until they push "Start" on the review "home" page. That's how we've implemented it so far for Reviews, so keep the same logic for _all_ reviews?

> This links a little bit to the comments above. How does the assigner know that all of the application has been assigned (just by looking at review_availability ?).

Yes, the idea is that this table would provide a quick look-up to show the assigned status of any relevant application. The logic that computes the table records would take into account what sections have been assigned. Essentially, if there are review_availability records that have an "available" value (as per enum above), then that application still has responses that haven't been assigned.

> There is a mention of level, what do you mean by that ?

Sorry, I mean "review level". In case it wasn't already clear, the review_availability table would only have records for the _next_ level's reviews. We don't compute subsequent level review availability until it's required (i.e. when at least one previous-level review is submitted -- is that right?)

> There is a bit of extra logic that needs to go to /review and /consolidation url where previously we just find the latest review, now we would need to find the latest review or review_availabity for the current stage. So we would need to amend that diagram.

Yes, true. I have a couple of thoughts about the URL pattern based on this discussion and other considerations, but I'll save that to a separate discussion.

**Comment by andreievg** [comment]
From our discussion:

**review_assignment** keep the name but have a field '**status**':
* **available** -> awaiting assignment by assigner
* **available_for_self_assignment** -> can be assigned by reviewer (see below)
* **non_available** -> can't be assigned anymore
* **assigned** -> assigned and means review/consolidation can be created

For any lvl of review, through trigger on either `review submitted` or `application submitted` we agreed that `review_assignments` will be created for lvl + 1 (including a new fields available_sections, an array of section ids). This doesn't quite help in figuring out if all have been assigned, but would help front end know what can be assigned to what reviewer (so from the front end perspective a GraphQL query to review_assignment is all assigner would need to figure out what can be assigned to whom).

**How do we know if application needs assignment ?**

Since this is only needed in the application list, we can add a column to the list view for this. Basically get count of all review_question_assignment distinct template_element ids, for review_assignments of lvl 1, and compare with count of template_elements of the application. (would be a boolean, something like `is_fully_assigned`).

**available_for_self_assignment** as mentioned above, means a reviewer can self assign (this is default for lvl + 1, but can also exist on lvl 1).  assigner_id on review_assignment will stay as NULL until it's assigned (will inherit either self assigner id or assigner id). Can restrict this editing with permissions btw (can check reviewer_id for self assignment, and the assigner is able to edit any assignments for permission they have based on permission restriction lvl/stage/template).

**What to do if reviewer is added**

This is a likely scenario, where there is a large number of incoming applications and say a consultant is recruited to help with reviews. They might get permission to review things that are ongoing (and that already have review_assignments made). We are parking this for now but did see a way to do this through an action/end point that updates review_assignments of current lvl of applications in progress (that haven't been assigned yet).

**Reviewer and Consolidator at the same time**

For say screening, where reviewer and consolidator is the same person, we can present 'overall decision' component to them. It seems the easiest way to do this is with permission restriction, 'canMakeOverallDecision' -> which will be recorded against review_assignment when it's created (can include as computer/generate column in review table). This would mean reviewer would be able to choose an overall decision. (and for forms like join company and screening, workflow would be quite easy)

**Quick explanation of what list shows for reviewer**

For reviewer at any lvl. 

Query review_assignmnet = 'assigned' or 'available for self assignment' for current application  (matching reviewer_id to reviewer id).

If latest reviewer_assignment = 'assigned' and review does not exist -> show can start/create
If latest reviewer_assignment = 'assigned' and does exist -> (based on status of review no, as per before)
If latest reviewer_assignment = 'available for self assignment' -> show can start/create

Above actual actions of start/create or continue happens through start page.

**URL**

Had a little discussion about URL, agreed for now to have both consolidation and review pages as just review (this diagram shows they are ver similar `https://app.diagrams.net/#G1IlIqhhKkRhc8tvudThjqr7uWdgHFpqTs`.

For now won't be breaking down URL into stages etc (also realised that stage names etc.. will be localised most likely, so not sure if that would work in URL).

Actionables
* update schema diagram
* update flow and lifecycle diagrams
* add `is_fully_assigned` to application list view
